test/e2e/clusterctl_upgrade.go

```diff
@@ -345,15 +352,46 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			client.MatchingLabels{clusterv1.ClusterLabelName: workLoadClusterName},
 		)
 		Expect(err).NotTo(HaveOccurred())
-
-		By("Upgrading providers to the latest version available")
-		clusterctl.UpgradeManagementClusterAndWait(ctx, clusterctl.UpgradeManagementClusterAndWaitInput{
-			ClusterctlConfigPath: input.ClusterctlConfigPath,
-			ClusterctlVariables:  input.UpgradeClusterctlVariables,
-			ClusterProxy:         managementClusterProxy,
-			Contract:             clusterv1.GroupVersion.Version,
-			LogFolder:            filepath.Join(input.ArtifactFolder, "clusters", cluster.Name),
-		}, input.E2EConfig.GetIntervals(specName, "wait-controllers")...)
+		// Check if the user want a custom upgrade
+		isCustomUpgrade := input.CoreProvider != "" ||
+			len(input.BootstrapProviders) > 0 ||
+			len(input.ControlPlaneProviders) > 0 ||
+			len(input.InfrastructureProviders) > 0 ||
+			len(input.IPAMProviders) > 0 ||
+			len(input.RuntimeExtensionProviders) > 0
+
+		if isCustomUpgrade {
+			By("Upgrading providers to custom versions")
+			log.Logf("CoreProvider: %s BootstrapProviders: %s ControlPlaneProviders: %s InfrastructureProviders: %s IPAMProviders: %s RuntimeExtensionProviders: %s",
+				input.CoreProvider,
+				strings.Join(input.BootstrapProviders, ","),
+				strings.Join(input.ControlPlaneProviders, ","),
+				strings.Join(input.InfrastructureProviders, ","),
+				strings.Join(input.IPAMProviders, ","),
+				strings.Join(input.RuntimeExtensionProviders, ","),
+			)
```

Comment:
```suggestion
```
nit: I think we can drop this here given we log it in the Upgrade func.

Just to reduce noise